### PR TITLE
[onnx] Fix bug for exporting torch.cdist into onnx and support 'compute_mode'

### LINKED ISSUE #144212: [onnx] export unexpected onnx for torch.cdist, which is slow and easy to cause CUDA OOM
### 🐛 Describe the bug

In [cdist,](https://github.com/pytorch/pytorch/blob/main/torch/onnx/symbolic_opset9.py#L6181) the 'compute_mode' was ignored, which leads to a big difference of the computation flow between original torch.cdist and the exported onnx file when computing Euclidean distance (p=2). For computing Euclidean distance, the running of exported onnx model will be 10x slower than running torch.cdist directly, and also very likely to cause CUDA OOM for larger matrixes unnecessarily.

here is the code for replicate this:
```python

import torch
import onnxruntime as ort
import onnx
import torch.onnx
import numpy as np
import time
import ctypes
import onnxruntime as ort
# export LD_LIBRARY_PATH=/home/vipuser/miniconda3/envs/torch/lib/python3.12/site-packages/nvidia/cudnn/lib/:$LD_CLIBRARY_PATH

class Averager:
    def __init__(self):
        self.call_count = {}
        self.total_sum = {}    

    def __call__(self, name, num):
        if name not in self.call_count:
            self.call_count[name] = 0
        if name not in self.total_sum:
            self.total_sum[name] = 0
            # ignore first time
            return
        self.call_count[name] += 1                # Increment call count
        self.total_sum[name] += num            # Add result to total sum
    
    def get(self, name):
        assert name in self.call_count
        assert name in self.total_sum
        return self.total_sum[name]/self.call_count[name]


class Cdist(torch.nn.Module):
    def forward(self, x, y, p=2.0, compute_mode="use_mm_for_euclid_dist_if_necessary"):
        # type: (Tensor, Tensor, float, str) -> (Tensor)
        r"""Computes batched the p-norm distance between each pair of the two collections of row vectors.

        Args:
            x1 (Tensor): input tensor of shape :math:`B \times P \times M`.
            x2 (Tensor): input tensor of shape :math:`B \times R \times M`.
            p: p value for the p-norm distance to calculate between each vector pair
                :math:`\in [0, \infty]`.
            compute_mode:
                'use_mm_for_euclid_dist_if_necessary' - will use matrix multiplication approach to calculate
                euclidean distance (p = 2) if P > 25 or R > 25
                'use_mm_for_euclid_dist' - will always use matrix multiplication approach to calculate
                euclidean distance (p = 2)
                'donot_use_mm_for_euclid_dist' - will never use matrix multiplication approach to calculate
                euclidean distance (p = 2)
                Default: use_mm_for_euclid_dist_if_necessary.

        If x1 has shape :math:`B \times P \times M` and x2 has shape :math:`B \times R \times M` then the
        output will have shape :math:`B \times P \times R`.

        This function is equivalent to `scipy.spatial.distance.cdist(input,'minkowski', p=p)`
        if :math:`p \in (0, \infty)`. When :math:`p = 0` it is equivalent to
        `scipy.spatial.distance.cdist(input, 'hamming') * M`. When :math:`p = \infty`, the closest
        scipy function is `scipy.spatial.distance.cdist(xn, lambda x, y: np.abs(x - y).max())`.

        """
        return torch.cdist(x, y, p, compute_mode)
    

def test(compute_mode = "use_mm_for_euclid_dist", p=2.0):
    print(">>", compute_mode, p)
    model = Cdist()  # Instantiate the model
    # Export the model to ONNX format
    x = torch.randn(1, 640, 128, ) 
    y = torch.randn(2, 1280, 128, )

    torch.onnx.export(model,
                    (x, y, p, compute_mode),  # Pass the two dummy inputs as a tuple
                    "cdist_.onnx",  # Output ONNX file
                    input_names=["x", "y",],  # Names for the input tensors
                    output_names=["output"],  # Name for the output tensor
                    #   dynamic_axes={"input1": {0: "batch_size"}, "input2": {0: "batch_size"}, "output": {0: "batch_size"}},  # Optional: dynamic batch size
                    opset_version=9)  # Specify the ONNX opset version

    inputs = [x, y]
    names = ["x", "y"]
    inputs_dict = {i:j for i, j in zip(names, inputs)}
    
    start = time.time()
    output_pytorch = model(x.cuda(), y.cuda(), p, compute_mode).cpu()
    # print(output_pytorch.size(), output_pytorch[0][0][:3])
    # print("torch model run for:", time.time()-start, "s")
    

    inputs_dict = {i:j.numpy() for i, j in zip(names, inputs)}
    onnx_model_path = "cdist_.onnx"  # Replace with the path to your ONNX model
    session = ort.InferenceSession(onnx_model_path, providers=['CUDAExecutionProvider'])
    start = time.time()
    onnx_output = session.run(["output"], inputs_dict)[0]
    onnx_time_cost = time.time()-start
    # print("onnx model run for:", onnx_time_cost, "s")
    onnx_output_tensor = torch.from_numpy(onnx_output)
    # print(onnx_output_tensor.size(), onnx_output_tensor[0][0][:3])
    if not torch.allclose(output_pytorch, onnx_output_tensor, rtol=1e-2):
        print("The outputs of the PyTorch model and the ONNX model are different!")
    else:
        print("ok")
    return onnx_time_cost

if __name__=="__main__":
    avg = Averager()
    testtimes = 20
    p = 2.0
    for i in range(testtimes):
        avg("donot_use_mm_for_euclid_dist", test("donot_use_mm_for_euclid_dist", p))
        avg("use_mm_for_euclid_dist_if_necessary", test("use_mm_for_euclid_dist_if_necessary", p))
        avg("use_mm_for_euclid_dist", test("use_mm_for_euclid_dist", p))
    print("donot_use_mm_for_euclid_dist: average time:", avg.get("donot_use_mm_for_euclid_dist"))
    print("use_mm_for_euclid_dist_if_necessary: average time:", avg.get("use_mm_for_euclid_dist_if_necessary"))
    print("use_mm_for_euclid_dist: average time:", avg.get("use_mm_for_euclid_dist"))


```
Just compare the time cost between torch.cdist and exported cdist model at different mode will show the difference. 
here is a result of the difference between three mode for the exported onnx model:
#### before fixing the bug, the time cost for onnx model is all the same. 
```
donot_use_mm_for_euclid_dist: average time: 0.09469890594482422
use_mm_for_euclid_dist_if_necessary: average time: 0.08542945510462711
use_mm_for_euclid_dist: average time: 0.09019689810903449
```
#### after fixing the bug,  the time cost for onnx model is 10x faster when using the accelerated computation method.
```
donot_use_mm_for_euclid_dist: average time: 0.10753977926153886
use_mm_for_euclid_dist_if_necessary: average time: 0.008870099720201995
use_mm_for_euclid_dist: average time: 0.011054039001464844
```

### Versions

PyTorch version: 2.5.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.12.7 | packaged by Anaconda, Inc. | (main, Oct  4 2024, 13:27:36) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.2.0-37-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.4.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080
Nvidia driver version: 550.127.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      40 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             8
On-line CPU(s) list:                0-7
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Gold 6133 CPU @ 2.50GHz
CPU family:                         6
Model:                              85
Thread(s) per core:                 2
Core(s) per socket:                 4
Socket(s):                          1
Stepping:                           4
BogoMIPS:                           4999.90
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 arat pku ospke md_clear arch_capabilities
L1d cache:                          256 KiB (8 instances)
L1i cache:                          256 KiB (8 instances)
L2 cache:                           16 MiB (4 instances)
L3 cache:                           16 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-7
Vulnerability Gather data sampling: Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Mitigation; PTE Inversion
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:             Mitigation; IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; IBRS, IBPB conditional, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; Clear CPU buffers; SMT Host state unknown

Versions of relevant libraries:
[pip3] numpy==2.1.3
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] onnx==1.17.0
[pip3] onnxruntime-gpu==1.20.1
[pip3] torch==2.5.0
[pip3] torchaudio==2.5.0
[pip3] torchvision==0.20.0
[pip3] triton==3.1.0
[conda] numpy                     2.1.3                    pypi_0    pypi
[conda] nvidia-cublas-cu12        12.4.5.8                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.4.127                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.2.1.3                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.5.147               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.6.1.9                 pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.3.1.170               pypi_0    pypi
[conda] nvidia-nccl-cu12          2.21.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.4.127                 pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.4.127                 pypi_0    pypi
[conda] torch                     2.5.0                    pypi_0    pypi
[conda] torchaudio                2.5.0                    pypi_0    pypi
[conda] torchvision               0.20.0                   pypi_0    pypi
[conda] triton                    3.1.0                    pypi_0    pypi

### PR DESCRIPTION
**Pull request by ygq65536**
### Fix bug for exporting torch.cdist and support 'compute_mode' 
In [cdist,](https://github.com/pytorch/pytorch/blob/main/torch/onnx/symbolic_opset9.py#L6181) the 'compute_mode' was ignored, which leads to a big difference of the computation flow between original torch.cdist and the exported onnx file when computing Euclidean distance (p=2). For computing Euclidean distance, the running of exported onnx model will be 10x slower than running torch.cdist directly, and also very likely to cause CUDA OOM for larger matrixes unnecessarily.

This code is going for exporting the same onnx computation flow with the forward of  torch.cdist defined at [forward implementation](https://github.com/pytorch/pytorch/blob/9225f149ebec1ff16d0ef31105ee8ecf4fb09efc/aten/src/ATen/native/Distance.cpp#L66-L149.) under every compute_mode.

Fixes #144212

### CODE REVIEW

File: torch/onnx/symbolic_opset9.py [hunk 1]
```diff
@@ -6190,8 +6190,17 @@ def cdist(
...
-    # using matrix multiplication to calculate the euclidean distance
+    row_size_x1 = symbolic_helper._get_tensor_dim_size(x1, -2)
+    row_size_x2 = symbolic_helper._get_tensor_dim_size(x2, -2)
+    assert row_size_x1 is not None
+    assert row_size_x2 is not None
+    p_float = symbolic_helper._parse_arg(p, "f")
+    compute_mode = symbolic_helper._parse_arg(compute_mode, "i")
+    if p_float == 2.0 and (
+        compute_mode == 1
+        or (compute_mode is None and row_size_x1 >= 25 and row_size_x2 >= 25)
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset9.py:6201]: Could you explain the size>=25 condition? Thanks

Comment (ygq65536) [reply torch/onnx/symbolic_opset9.py:6201]: This condition is aligning with the condition used in the original implementation and the documentation.
implementation for torch.cdist: https://github.com/pytorch/pytorch/blob/9225f149ebec1ff16d0ef31105ee8ecf4fb09efc/aten/src/ATen/native/Distance.cpp#L137 
document for torch.cdist:
https://pytorch.org/docs/stable/generated/torch.cdist.html

**Comment by ygq65536** [comment]
this is my first time starting a PR for pytorch. Could someone please tell me whether the CI status is correct. Because I just expect there should be more test results, but I got only a few results like lint and label check, nothing else.
image

**Comment by ygq65536** [comment]
@justinchuby  Good Morning, Justin. The CI test failure of this PR in [test_fp8_matmul1;](https://hud.pytorch.org/pr/144213) I think it is irrelevant, can you help me skip it? Or there is a different solution, please tell me.

**Review by justinchuby — APPROVED** [review]
Thank you!
